internal/server/v2/observation/calculation_util.go

```diff
@@ -191,3 +195,105 @@ func findObservationResponseHoles(
 	}
 	return result, nil
 }
+
+// Find all candidate observations that match each ASTNode and add to VariableFormula.
+func computeLeafObs(
+	inputResp *pbv2.ObservationResponse,
+	formula *VariableFormula,
+) {
+	for _, leafData := range formula.LeafData {
+		leafData.CandidateObs = map[string]map[string][]*pb.PointStat{}
+		variableObs, ok := inputResp.ByVariable[leafData.StatVar]
+		// No data for input variable.
+		if !ok {
+			return
+		}
+		for entity, entityObs := range variableObs.ByEntity {
+			facetMap := map[string][]*pb.PointStat{}
+			for _, facetObs := range entityObs.OrderedFacets {
+				if leafData.Facet != nil {
+					facet := inputResp.Facets[facetObs.FacetId]
+					if leafData.Facet.GetMeasurementMethod() != facet.GetMeasurementMethod() {
+						continue
+					}
+					if leafData.Facet.GetObservationPeriod() != facet.GetObservationPeriod() {
+						continue
+					}
+					if leafData.Facet.GetUnit() != facet.GetUnit() {
+						continue
+					}
+					if leafData.Facet.GetScalingFactor() != facet.GetScalingFactor() {
+						continue
+					}
+				}
+				facetMap[facetObs.FacetId] = facetObs.Observations
+			}
+			if len(facetMap) > 0 {
+				leafData.CandidateObs[entity] = facetMap
+			}
+		}
+	}
+}
+
+// Combine two sets of candidate observations in one step of calculation.
+func evalBinaryExpr(
+	x, y map[string]map[string][]*pb.PointStat,
+	op token.Token,
+) (map[string]map[string][]*pb.PointStat, error) {
+	result := map[string]map[string][]*pb.PointStat{}
+	for entity, xFacetObs := range x {
+		newFacetObs := map[string][]*pb.PointStat{}
+		for facetId, xObs := range xFacetObs {
+			yFacetObs, ok := y[entity]
+			if !ok {
+				continue
+			}
+			yObs, ok := yFacetObs[facetId]
+			if !ok {
+				continue
+			}
+			newObs := []*pb.PointStat{}
+			xIdx, yIdx := 0, 0
+			for xIdx < len(xObs) {
+				if yIdx == len(yObs) {
+					break
+				}
+				if xObs[xIdx].GetDate() == yObs[yIdx].GetDate() {
```

Comment:
I am not fully sure the date alignment all works. Can you add some test case for non uniform and partially aligned date cases?

Reply:
Good point, I added some more tests and expanded the main alignment test and changed the alignment to follow https://takeuforward.org/data-structure/intersection-of-two-sorted-arrays/ (which seems to be the standard way to do it)